python/src/iceberg/table/partitioning.py

```diff
@@ -14,8 +14,13 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+from typing import Dict, Iterable, List, Tuple
+
+from iceberg.schema import Schema
 from iceberg.transforms import Transform
 
+_PARTITION_DATA_ID_START: int = 1000
+
 
 class PartitionField:
```

Comment:
Instead of implementing the `__eq__` and `__hash__` we could leverage the dataclass library. If we set `eq=True` and `frozen=True` (which makes it immutable, which is also nice), then we get __hash__ automatically:

If `eq` and `frozen` are both true, by default [dataclass()](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) will generate a `__hash__()` method for you. If eq is true and frozen is false, `__hash__()` will be set to `None`, marking it unhashable (which it is, since it is mutable). If `eq` is false, `__hash__()` will be left untouched meaning the `__hash__()` method of the superclass will be used (if the superclass is [object](https://docs.python.org/3/library/functions.html#object), this means it will fall back to id-based hashing).

More information here: https://docs.python.org/3/library/dataclasses.html

Reply:
yeah I agree the PartitionField is an immutable class after construction so dataclass with both eq and frozen sounds fair to me. 

On the other side, I think the biggest benefit of the dataclass is the `__post_init__` method which allow for java-like builderPattern equivalent processing when we build the PartitionSpec. There's collection of validations need to happen and I am discussing with @samredai  in https://github.com/apache/iceberg/issues/4631#issuecomment-1113632408. 

From what I can tell, we will need a `PartitionSpecBuilder` class with convenient way to construct the PartitionSpec, but we also want to make sure avoid duplicate the builder logic in an overly complex init method for PartitionSpec

Reply:
@dramaticlly that's an interesting idea I haven't thought of. A big argument for using the builder pattern was that we wanted `PartitionSpec` to be immutable, which would require us to include a ton of validation logic (everything that would be in a builder) in the `__init__` method. If I understand your suggestion, using `__post_init__` would allow us to have a typical init method, but then include the builder-type validation logic in the `__post_init__` which would fail the initialization of any invalid `PartitionSpec`.

Reply:
Is the plan to use `@dataclass`? I like that idea, but I won't hold off on reviewing if we want to get it in like this first.

Reply:
I really like this idea of wrapping the classes generated from the OpenAPI spec. The class naming here might be tricky. Module namespacing allows us to re-use the same name if we want, something like:
```py
from iceberg.openapi import rest_catalog

class PartitionField(rest_catalog.PartitionField):
    ...
```
We shouldn't expect users to import from the openapi module directly so we shouldn't need to worry about naming conflicts, right? Maybe we should name it `_openapi` just to be super clear about that.

Reply:
I'll hold off on commenting too much until I have a chance to look into the `pydantic` project as well as look at the other PR.

My first ask would be how many dependencies are we bringing in if we add `pydantic`? I know that some folks were concerned about adding very many external python dependencies, so as not to conflict with their own dependencies, but if the benefit is very large I'm not personally opposed to it (I believe it was somebody / some group from Netflix that originally requested we keep the number of required dependencies down).

But validation, either via a library or via a common pattern we settle on, is something that would be very beneficial.

Reply:
Great point, I just checked the [dependencies](https://github.com/samuelcolvin/pydantic/blob/master/setup.py#L132) for pydantic and the good news is that all it requires is `typing-extensions` which is probably just for some python 3.7 backports. That will even probably get dropped at some point when they no longer support 3.7

Reply:
I would not consider this blocking for now.